rfcs/0000-rlp-v2.md

```diff
@@ -0,0 +1,740 @@
+
+To keep compatibility with the [Envoy Rate Limit protocol](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter), the well-known data structure can optionally be extended with the `RateLimitRequest`, thus resulting in the following final structure.
+
+```yaml
+context: # Envoy's Ext-Authz `CheckRequest.AttributeContext` type
+  source:
+    address: …
+    service: …
+    …
+  destination:
+    address: …
+    service: …
+    …
+  request:
+    http:
+      host: …
+      path: …
+      method: …
+      headers: {…}
+
+  auth: # Dynamic metadata exported by the external authorization service
+
+  ratelimit: # Envoy's Rate Limit `RateLimitRequest` type
+    domain: … # generated by the Kuadrant controller
+    descriptors: {…} # descriptors configured by the user directly in the proxy (not generated by the Kuadrant controller, if allowed)
+    hitsAddend: … # only in case we want to allow users to refer to this value in a policy
+```
+
+### Mechanics of generating RL descriptor actions
+
+From the perspective of a user who writes a RLP, the selectors used in then `when` and `counters` fields are paths to the well-known data structure ([_well-known selectors_](#well-known-selectors)). While desiging a policy, the user intuitively pictures the well-known data structure and designs each rule (each limit) thinking on the possible values assumed by each of those paths in the data plane. For example,
+
+The user story:
+> _Whenever the context is an HTTP request sent to `dolls.toystore.com` (`context.request.http.host`), I want a rate limit of 50 rps per distinct user (`auth.identity.username`)._
+
+Materializes as the following RLP:
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: toys-path
+      when:
+        - selector: context.request.http.host
+          operator: eq
+          value: dolls.toystore.com
+      rates:
+        - limit: 50
+          duration: 1
+          unit: second
+      counters:
+        - auth.identity.username
+```
+
+The following selectors are to be interpreted:
+- `context.request.http.host`
+- `auth.identity.username`
+
+The RLP controller uses a map to translate each selector into its corresponding descriptor action. Roughly described:
+
+```
+context.source.address    → source_cluster(...) # TBC
```

Comment:
why using envoy RL filter configuration directives? Why not implement some more flexible and generic "language" to be able to use "operators" for instance? It is all about creating descriptors based on rules. For instance, the wasm module might be able to understand

```
        - selector: context.request.http.host
          operator: eq
          value: dolls.toystore.com
```

And add descriptor entry when the selector matches.

Reply:
Please see https://github.com/Kuadrant/architecture/pull/8#discussion_r1118655040.